[DAGCombiner] Add DAG combine for PARTIAL_REDUCE_MLA when no mul op

NickGuy-Arm opened the pull request:
Generic DAG combine for ISD::PARTIAL_REDUCE_U/SMLA to convert: PARTIAL_REDUCE_*MLA(Acc, ZEXT(UnextOp1), Splat(1)) into PARTIAL_REDUCE_UMLA(Acc, UnextOp1, TRUNC(Splat(1))) and
PARTIAL_REDUCE_*MLA(Acc, SEXT(UnextOp1), Splat(1)) into PARTIAL_REDUCE_SMLA(Acc, UnextOp1, TRUNC(Splat(1))).

@JamesChesterman is the original author.

NickGuy-Arm commented:
This PR implicitly depends on https://github.com/llvm/llvm-project/pull/130935

sdesmalen-arm commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -12669,6 +12679,48 @@ SDValue DAGCombiner::visitPARTIAL_REDUCE_MLA(SDNode *N) {
                      RHSExtOp);
 }
 
+// Makes PARTIAL_REDUCE_*MLA(Acc, ZEXT(UnextOp1), Splat(1)) into
+// PARTIAL_REDUCE_UMLA(Acc, Op, TRUNC(Splat(1)))
+// Makes PARTIAL_REDUCE_*MLA(Acc, SEXT(UnextOp1), Splat(1)) into
+// PARTIAL_REDUCE_SMLA(Acc, Op, TRUNC(Splat(1)))
```
```suggestion
// partial.reduce.umla(acc, zext(op1), splat(s))
// -> partial.reduce.umla(acc, op1, splat(trunc(s))
// partial.reduce.smla(acc, sext(op1), splat(s))
// -> partial.reduce.smla(acc, op1, splat(trunc(s))
```

sdesmalen-arm commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -12669,6 +12679,48 @@ SDValue DAGCombiner::visitPARTIAL_REDUCE_MLA(SDNode *N) {
+  SDValue Op2 = N->getOperand(2);
+
+  APInt ConstantOne;
+  if (!ISD::isConstantSplatVector(Op2.getNode(), ConstantOne) ||
+      !ConstantOne.isOne())
+    return SDValue();
```
We can do this for all values that fit in the smaller type, not just the constant `1`. i.e. if we know that `zext(trunc(op2)) == op2` if op1 is zero-extended  or `sext(trunc(op2)) == op2` if op1 is sign-extended, then we can do this transform. We could use `ComputeNumSignBits` to determine if we can truncate the sign bits.

NickGuy-Arm replied:
I've looked into this and, as far as I can tell, we can't do this. I guess more specifically there is no value in it, as the only cases where this code path is executed, `Op2` is always a constant `1`.
Trying to manipulate this from IR by making the mul operand also have a constant operand will prevent said mul from being folded into the mla altogether, as it no longer has two extend operands.
The only way I can see to achieve this is to loosen the requirement of both operands of a mul needing to be extended, but that then risks discarding other semantics (such as the no-wrap flags, as you've pointed out offline).

If we do want to loosen the requirements, then I think that would be better suited in a distinct PR. But within the context of this PR, I would say this is not something we can do.

NickGuy-Arm committed:
Update comment

NickGuy-Arm committed:
Update tests after rebase

sdesmalen-arm commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -12669,6 +12679,48 @@ SDValue DAGCombiner::visitPARTIAL_REDUCE_MLA(SDNode *N) {
+  EVT AccElemVT = Acc.getValueType().getVectorElementType();
+  if (Op1IsSigned != NodeIsSigned &&
+      (Op1.getValueType().getVectorElementType() != AccElemVT ||
+       Op2.getValueType().getVectorElementType() != AccElemVT))
```
I don't think you need to test the `Op2` case here, because the type of `Op1` must match that of `Op2`.

sdesmalen-arm commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -12669,6 +12679,48 @@ SDValue DAGCombiner::visitPARTIAL_REDUCE_MLA(SDNode *N) {
+  if (!TLI.isPartialReduceMLALegalOrCustom(N->getValueType(0), UnextOp1VT))
+    return SDValue();
+
+  SDValue TruncOp2 = DAG.getNode(ISD::TRUNCATE, DL, UnextOp1VT, Op2);
```
This can just be `DAG.getConstant(1, DL, UnextOp1VT)` (and better to just inline that in the use below).

sdesmalen-arm commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -12669,6 +12679,48 @@ SDValue DAGCombiner::visitPARTIAL_REDUCE_MLA(SDNode *N) {
+
+  SDValue UnextOp1 = Op1.getOperand(0);
+  EVT UnextOp1VT = UnextOp1.getValueType();
+
```
nit: remove newline.

sdesmalen-arm commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -12669,6 +12679,48 @@ SDValue DAGCombiner::visitPARTIAL_REDUCE_MLA(SDNode *N) {
+  SDValue TruncOp2 = DAG.getNode(ISD::TRUNCATE, DL, UnextOp1VT, Op2);
+
+  bool Op1IsSigned = Op1Opcode == ISD::SIGN_EXTEND;
+
```
nit: remove newline.

sdesmalen-arm commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -12669,6 +12679,48 @@ SDValue DAGCombiner::visitPARTIAL_REDUCE_MLA(SDNode *N) {
+// partial.reduce.umla(acc, op, splat(trunc(1)))
+// Makes partial.reduce.smla(acc, sext(op1), splat(1)) into
+// partial.reduce.smla(acc, op, splat(trunc(1)))
+SDValue DAGCombiner::foldPartialReduceMLANoMulOp(SDNode *N) {
```
nit: maybe rename this to `foldPartialReduceAdd`?

sdesmalen-arm commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -12669,6 +12679,48 @@ SDValue DAGCombiner::visitPARTIAL_REDUCE_MLA(SDNode *N) {
                      RHSExtOp);
 }
 
+// Makes partial.reduce.umla(acc, zext(op1), splat(1)) into
+// partial.reduce.umla(acc, op, splat(trunc(1)))
+// Makes partial.reduce.smla(acc, sext(op1), splat(1)) into
+// partial.reduce.smla(acc, op, splat(trunc(1)))
```
nit:
```suggestion
// partial.reduce.umla(acc, zext(op1), splat(1))
// -> partial.reduce.umla(acc, op, splat(trunc(1)))
// partial.reduce.smla(acc, sext(op1), splat(1))
// -> partial.reduce.smla(acc, op, splat(trunc(1)))
```

sdesmalen-arm reviewed (commented):
Mostly some nits, otherwise looks good to me.

NickGuy-Arm committed:
Address nits

NickGuy-Arm committed:
[DAGCombiner] Add DAG combine for PARTIAL_REDUCE_MLA when no mul op

Generic DAG combine for ISD::PARTIAL_REDUCE_U/SMLA to convert:
PARTIAL_REDUCE_*MLA(Acc, ZEXT(UnextOp1), Splat(1)) into
PARTIAL_REDUCE_UMLA(Acc, UnextOp1, TRUNC(Splat(1)))
and
PARTIAL_REDUCE_*MLA(Acc, SEXT(UnextOp1), Splat(1)) into
PARTIAL_REDUCE_SMLA(Acc, UnextOp1, TRUNC(Splat(1))).
